File: LeetCode/369-plus-one-linked-list/ex_369.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def plusOne(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        prev = ListNode(0)
        prev.next = head
        p = head
        start = prev
        while p:
            if p.val != 9:
                start = p
            prev = p
            p = p.next
        if prev.val != 9:
            prev.val += 1
            return head
        else:
            start.val += 1
            r = start.next
            while r:
                r.val = 0
                r = r.next
            return start if start.next == head else head

File: LeetCode/369-plus-one-linked-list/ex_369.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* plusOne(ListNode* head) {
        ListNode* prev = new ListNode(0);
        prev->next = head;
        ListNode *p = head, *start = prev;
        while (p != NULL) {
            if (p->val != 9)
                start = p;
            prev = p;
            p = p->next;
        }
        if (prev->val != 9) {
            prev->val += 1;
            return head;
        }
        else {
            ++start->val;
            ListNode *r = start->next;
            while (r != NULL) {
                r->val = 0;
                r = r->next;
            }
            return start->next == head ? start : head;
        }
    }
};